### PR TITLE
Add gpu.create_stream op and add stream as an optional argument in few gpu dialect ops

### PR DESCRIPTION
**Pull request by nbpatel**
# **DO NOT FILE A PULL REQUEST**

This repository does not accept pull requests. Please follow http://llvm.org/docs/Contributing.html#how-to-submit-a-patch for contribution to LLVM.

# **DO NOT FILE A PULL REQUEST**

### CODE REVIEW

File: mlir/test/Dialect/GPU/ops.mlir [hunk 1]
```diff
@@ -234,6 +241,12 @@ module attributes {gpu.container_module} {
     // CHECK: gpu.dealloc %[[m2]] : memref<13xf32, 1>
     gpu.dealloc %m2 : memref<13xf32, 1>
 
+    %stream = gpu.create_stream : !gpu.stream
```
Comment (Hardcode84) [inline mlir/test/Dialect/GPU/ops.mlir:246]: `// CHECK %[[stream:.*]] = gpu.create_stream : !gpu.stream`

File: mlir/test/Dialect/GPU/ops.mlir [hunk 2]
```diff
@@ -234,6 +241,12 @@ module attributes {gpu.container_module} {
...
+    // CHECK: %[[m3:.*]] = gpu.alloc () stream %[[stream:.*]] : memref<13xf32, 1>
```
Comment (Hardcode84) [inline mlir/test/Dialect/GPU/ops.mlir]: `%[[stream]]`

File: mlir/lib/Dialect/GPU/IR/GPUDialect.cpp [hunk 3]
```diff
@@ -164,6 +166,13 @@ Type GPUDialect::parseType(DialectAsmParser &parser) const {
   if (keyword == "async.token")
     return AsyncTokenType::get(context);
 
+  if (keyword == "stream") {
```
Comment (Hardcode84) [inline mlir/lib/Dialect/GPU/IR/GPUDialect.cpp]: remove curly braces

File: mlir/lib/Dialect/GPU/IR/GPUDialect.cpp [hunk 4]
```diff
@@ -1327,17 +1377,22 @@ struct EraseRedundantGpuWaitOpPairs : public OpRewritePattern<WaitOp> {
                                 PatternRewriter &rewriter) const final {
     auto predicate = [](Value value) {
       auto waitOp = value.getDefiningOp<WaitOp>();
-      return waitOp && waitOp->getNumOperands() == 0;
+      return waitOp && waitOp.getAsyncDependencies().size() == 0;
     };
-    if (llvm::none_of(op.getAsyncDependencies(), predicate))
+
+    if (llvm::none_of(op.getAsyncDependencies(), predicate)) {
```
Comment (Hardcode84) [inline mlir/lib/Dialect/GPU/IR/GPUDialect.cpp]: remove curly braces on single-line ifs

**Review by Hardcode84 — APPROVED** [review]

**Review by silee2 — APPROVED** [review]
LGTM

**Pull request closed without merging**